server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -410,6 +441,200 @@ public void clusterChanged(ClusterChangedEvent event) {
                 }
             }
         }
+        if (currentMaster == null && clusterService.localNode().isMasterNode()) {
+            /*
+             * This begins polling all master-eligible nodes for cluster formation information. However there's a 10-second delay before it
+             * starts, so in the normal situation where during a master transition it flips from master1 -> null -> master2, it the
+             * polling tasks will be canceled before any requests are actually made.
+             */
+            beginPollingClusterFormationInfo();
+        } else {
+            cancelPollingClusterFormationInfo();
+        }
+    }
+
+    private void beginPollingClusterFormationInfo() {
+        cancelPollingClusterFormationInfo();
+        clusterFormationInfoTasks = getMasterEligibleNodes().stream()
+            .map(masterNode -> beginPollingClusterFormationInfo(masterNode, clusterFormationResponses))
+            .collect(Collectors.toList());
+    }
+
+    private void cancelPollingClusterFormationInfo() {
+        clusterFormationInfoTasks.forEach(Scheduler.Cancellable::cancel);
+        /*
+         * Recreates the map so that we don't read old information, or worse get stuck with information about a node that has been
+         * removed from the cluster.
+         */
+        clusterFormationResponses = new ConcurrentHashMap<>();
+    }
+
+    Scheduler.Cancellable beginPollingClusterFormationInfo(
+        // Non-private for testing
+        DiscoveryNode node,
+        final ConcurrentMap<DiscoveryNode, ClusterFormationStateOrException> nodeToClusterFormationStateMap
+    ) {
+        return new PollClusterFormationStateTask(node, nodeToClusterFormationStateMap).pollUntilCancelled();
+    }
+
+    /**
+     * This class represents a collection of related Cancellables. If one is cancelled, they are all considered cancelled. If cancel() is
+     * called on this method, then cancel() is called on all child Cancellables.
+     */
+    private static class MultipleCancellablesWrapper implements Scheduler.Cancellable {
+        private final List<Scheduler.Cancellable> delegates;
+
+        MultipleCancellablesWrapper() {
+            this.delegates = Collections.synchronizedList(new ArrayList<>());
```

Comment:
Ignore this for now. As it is, I think it could blow up when we're iterating through this list. I'm not fixing it for the moment since I've sent out an email with line numbers.